homeassistant/components/smarttub/sensor.py

```diff
@@ -2,17 +2,48 @@
 from enum import Enum
 import logging
 
+import smarttub
+import voluptuous as vol
+
+from homeassistant.const import ATTR_ENTITY_ID
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers import config_validation as cv, entity_platform
+
 from .const import DOMAIN, SMARTTUB_CONTROLLER
 from .entity import SmartTubSensorBase
 
 _LOGGER = logging.getLogger(__name__)
 
+
 ATTR_DURATION = "duration"
 ATTR_LAST_UPDATED = "last_updated"
 ATTR_MODE = "mode"
 ATTR_START_HOUR = "start_hour"
 
 
+SET_PRIMARY_FILTRATION_SCHEMA = vol.Schema(
+    {
+        vol.Required(ATTR_ENTITY_ID): cv.entity_id,
+        # TODO: at least one of these two should be required
+        vol.Optional(ATTR_DURATION): vol.All(int, vol.Range(min=1, max=24)),
+        vol.Optional(ATTR_START_HOUR): vol.All(int, vol.Range(min=0, max=23)),
+    }
+)
+
+SET_SECONDARY_FILTRATION_SCHEMA = vol.Schema(
+    {
+        vol.Required(ATTR_ENTITY_ID): cv.entity_id,
+        vol.Required(ATTR_MODE): vol.Any(
+            str,
```

Comment:
Is any string allowed here ?

Reply:
No, I think I had `Any` where I meant `All`

Reply:
It must be one of the values in the list which follows (derived from `SecondaryFiltrationMode`)

Reply:
Should it be like this?

```
SET_SECONDARY_FILTRATION_SCHEMA = vol.Schema(                                                                                                                                                   
    {                                                                                                                                                                                           
        vol.Required(ATTR_ENTITY_ID): cv.entity_id,                                                                                                                                             
        vol.Required(ATTR_MODE): [                                                                                                                                                              
                mode.name.lower()                                                                                                                                                               
                for mode in smarttub.SpaSecondaryFiltrationCycle.SecondaryFiltrationMode                                                                                                        
            ],                                                                                                                                                                                  
    }                                                                                                                                                                                           
)                                                                                                                                                                                               
```

that breaks with:

```
FAILED test_sensor.py::test_filtration_cycles[pyloop] - voluptuous.error.MultipleInvalid: expected a list for dictionary value @ data['mode']
```

Reply:
Sorry, I'm finding the voluptuous docs a bit difficult to decipher. I'll look for an example in another integration

Reply:
OK I think `vol.In` is what i need, will fix